### ISSUE TITLE
False Positive for too-many-function-args when subclassing __new__

### ISSUE THREAD

**Issue by Redoubts** [PyCQA/pylint#10670]
### Bug description

I subclass `datetime.datetime` for some specific needs, but with the new 4.0 version of pylint, I see E1121 when I try the following:

```python
import datetime
from typing import Self


class MyDateTime(datetime.datetime):

    def __new__(
        cls,
        firstArg: object,
        month: int | None = None,
        day: int | None = None,
        hour: int = 0,
        minute: int = 0,
        second: int = 0,
        microsecond: int = 0,
    ) -> Self:
        return super().__new__(
            cls, firstArg, month, day, hour, minute, second, microsecond
        )
```

I think this is pretty valid (I tried to strip out as much as I could to repro), and this instantiates fine

```
In [3]: MyDateTime(1,1,1)
Out[3]: MyDateTime(1, 1, 1, 0, 0)
```

Am I misreading the issue, or is something off with pylint's interpretation

### Configuration

```ini

```

### Command used

```shell
pylint --rcfile /dev/null --disable all --enable E1121 x.py
```

### Pylint output

```python
************* Module x
x.py:17:15: E1121: Too many positional arguments for classmethod call (too-many-function-args)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 8.89/10, -8.89)
```

### Expected behavior

"10/10, would lint again"

### Pylint version

```shell
pylint 4.0.1
astroid 4.0.1
Python 3.11.13
```

### OS / Environment

Debian

### Additional dependencies

```python

```

**Comment by ekanshul** [comment]
This still reproduces on Python 3.11 with pylint 4.0.6 / astroid 4.0.4 (the snippet from the report gives `E1121` at the `super().__new__(...)` line). It stops reproducing on 3.12+ only because `datetime` resolves to `_pydatetime` there; the underlying problem is in astroid and also hits `array.array.__new__(cls, "b", data)` and `ssl.SSLSocket._create`'s `cls.__new__(cls, **kwargs)` on every version.

Root cause and fix: pylint-dev/astroid#3237 / pylint-dev/astroid#3238 — the fallback `__new__` that astroid hands out for C types outside `builtins` had the signature `(self, cls)`, while `__new__` takes the class explicitly, so every extra argument was "too many". With that branch installed the snippet is clean on 3.11.

Mentioning it because #11043 adds a `min_pyver=3.12` regression test for this issue; the test will pass, but the 3.11 case behind the report is a different fix.
